app/src/main/java/de/moyapro/idle/domain/skillTree/TreeOfLife.kt

```diff
@@ -2,40 +2,60 @@ package de.moyapro.idle.domain.skillTree
 
 import de.moyapro.idle.domain.traits.Feature
 
+class TreeOfLifeBuilder<T : Feature>(private val parent: T, private val tree: TreeOfLife<T>) {
+    fun branchInto(child: T, block: (TreeOfLifeBuilder<T>.() -> Unit)?) {
+        tree.add(parent, child)
+        block?.invoke(TreeOfLifeBuilder(child, tree))
+    }
+
+    fun branchInto(child: T) {
+        tree.add(parent, child)
+    }
+
+    fun create(): TreeOfLife<T> {
+        return tree
+    }
+}
+
 /**
  * TreeOfLife representing the large scale features in the game.
  * These Features are independent of each other only having other feature as requirements.
  * More refined evolution of more dependent features is implemented in bonsai tree with interdependent features
  */
-class TreeOfLife<T : Feature>(private var feature: T) {
-    private var children: MutableSet<TreeOfLife<T>> = mutableSetOf()
+class TreeOfLife<T : Feature> {
+    companion object {
+        fun <T : Feature> build(root: T, block: (TreeOfLifeBuilder<T>.() -> Unit)?): TreeOfLife<T> {
+            return TreeOfLifeBuilder(root, TreeOfLife()).let {
+                block?.invoke(it)
+                it
+            }.create()
+        }
+
+        fun <T : Feature> build(root: T): TreeOfLife<T> {
+            return TreeOfLifeBuilder(root, TreeOfLife()).create()
+        }
+    }
+
+    private var evolutions: MutableMap<T, MutableSet<T>> = mutableMapOf()
+
+    fun add(ancestor: T, descendants: MutableSet<T>): TreeOfLife<T> {
+        this.evolutions[ancestor] = descendants
+        return this
+    }
 
-    fun branchInto(feature: T): TreeOfLife<T> {
-        val child = TreeOfLife(feature)
-        children.add(child)
-        return child
+    fun add(ancestor: T, descendant: T): TreeOfLife<T> {
+        this.evolutions.getOrPut(ancestor, { mutableSetOf() }) += descendant
```

Comment:
Das ist ja nice. Ich will gar nicht wissen wie viele Zeilen das in Java waeren ^^